### PR TITLE
Port :filename.rootname/1 and :filename.rootname/2 to JS

### LINKED ISSUE #509: Port :filename.rootname/1 and :filename.rootname/2 to JS
### `:filename.rootname/1`

`-spec rootname(Filename) -> file:filename_all() when Filename :: file:name_all().`

### `:filename.rootname/2`

```
-spec rootname(Filename, Ext) -> file:filename_all()
                  when Filename :: file:name_all(), Ext :: file:name_all().
```

### PR DESCRIPTION
**Pull request by mward-sudo**
Closes #509 

## Summary by CodeRabbit

* **New Features**
  * Added rootname/1 and rootname/2 to remove file extensions from binary/text paths, with explicit-extension handling, improved input normalization (iolists/atoms), and UTF‑8-safe behavior for invalid encodings. Handles hidden files, trailing dots/slashes, and edge-case paths.

* **Tests**
  * Added extensive cross-language tests covering many path scenarios, extension matching/partial matches, unusual inputs, and argument-type/error cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: assets/js/erlang/filename.mjs [hunk 1]
```diff
@@ -49,6 +49,132 @@ const Erlang_Filename = {
...
+      if (!isAfterSlash) {
+        result = beforeExt;
+      }
+    }
+
+    // rootname/2 always returns a binary, regardless of input type
+    return Type.bitstring(result);
+  },
+  // End rootname/2
+  // Deps: [:erlang.iolist_to_binary/1]
```
[the remark this answers is not in the document]
Comment (mward-sudo) [reply assets/js/erlang/filename.mjs]: In the Elixir consistency tests, the following test passes:

`assert :filename.rootname([], ".erl") == ""`

However, asserting an empty list results in a failure:

`assert :filename.rootname([], ".erl") == []`

Also a failure: 

`assert :filename.rootname([], ".erl") == [""]`

This demonstrates that the above comment is not accurate, or at least inconsistent with the results we get from Elixir.

[the remark this answers is not in the document]
Comment (mward-sudo) [reply assets/js/erlang/filename.mjs]: I can confirm that running the following in erl does give an empty list.

```
1> filename:rootname([], ".erl"). 
[]
```

Should we maintain compatibility with the Elixir result or the Erlang result?

[the remark this answers is not in the document]
Comment (mward-sudo) [reply assets/js/erlang/filename.mjs]: As I see it, in the case of passing an empty list as the first argument to `:filename.rootname/2` we can:

- maintain consistency with Elixir by returning an empty binary

or

- strictly implement the Erlang spec and return an empty list, but lose consistency between the JS and Elixir tests as JS would test for an empty list and ExUnit for an empty binary

I do not know a way to coerce Elixir to return an empty list like Erlang does.

[the remark this answers is not in the document]
Comment (mward-sudo) [reply assets/js/erlang/filename.mjs]: @bartblast  do you agree with the AI conclusions here?

File: assets/js/erlang/filename.mjs [hunk 2]
```diff
@@ -102,6 +155,119 @@ const Erlang_Filename = {
...
+      filenameBinary.text === false || extBinary.text === false
+        ? processInvalidUtf8Extension(filenameBinary, extBinary)
+        : Type.bitstring(
+            removeIfNotAfterSlash(filenameBinary.text, extBinary.text),
+          );
+
+    return result;
+  },
+  // End rootname/2
+  // Deps: [:erlang.iolist_to_binary/1, :filename.flatten/1, :filename._rootname_raw/2]
```
[the remark this answers is not in the document]
Comment (mward-sudo) [reply assets/js/erlang/filename.mjs:659]: Elixir compatibility is the aim.

Elixir Path.rootname/2 uses :filename.rootname/2
Elixir Path.rootname/1 uses :filename.rootname/2

Could you check that the way Elixir uses the erlang functions makes your comment relevant.

**Comment by mward-sudo** [comment]
I've reviewed Elixir’s use of `:filename.rootname/1` and `:filename.rootname/2`.

The upshot of this review is that it only uses each Erlang function in `Path.rootname/1` and `Path.rootname/2` respectively.

Both functions in Elixir sanitise the parameter(s) by calling `IO.chardata_to_string(path)`. This guarantees that the Erlang function will only receive a valid UTF-8 binary.

As a result, I have simplified the implementation of this function to only do what Elixir requires of it. Should Elixir change its use, then we might need to expand the function to cover additional scenarios.

**Comment by mward-sudo** [comment]
@bartblast I'd appreciate some feedback on this approach to finding out how Elixir uses the Erlang functions before I progress some of my other work.

The main pro is to drop some complexity where Elixir doesn't require it, the major con is that we're only partially implementing the Erlang spec and Elixir could change the way it uses the function at any time.

I'd only take this approach if I thought the complexity savings are worthwhile. In the instance of the filename module, flattening file paths is a significant amount of code - but Elixir appears to interface with Erlang in a particular way that limits the types it uses and deep lists are not featured.

Thoughts welcome.

**Comment by bartblast** [comment]
@mward-sudo Thanks for digging into this. My take on it is that we should match Erlang behavior for Erlang stdlib functions, using the Erlang documentation as our specification.

The problem with only implementing the specific ways Elixir stdlib currently uses these functions is that we have no reliable way to track how Elixir uses them, and those usage patterns could change between Elixir versions. Even though Hologram advertises Elixir consistency, reverse-engineering from Elixir's current usage creates a moving target that could break without us detecting it.

Additionally, third-party Elixir libraries may call Erlang functions like `:filename.rootname/2` directly, expecting standard Erlang behavior. If we only support the patterns Elixir stdlib happens to use, these libraries could break in unexpected ways on the client-side.

So when Hologram claims to support an Erlang stdlib function, it should behave like the actual Erlang function, not just the subset of behaviors that Elixir happens to use today.

That said, I understand that some behaviors might not be practical or too much work at this stage. When that's the case, we should add a clear comment explaining what's not supported and raise a HologramInterpreterError for those specific inputs. At least this way we'll discover problems rather than having silent incorrect behavior where the function appears to work but doesn't match the actual Erlang behavior.

**Comment by mward-sudo** [comment]
Fully understood. I can recover the full implementation (sans OS specific handling which is inappropriate for web use).

Let me do this before you review this PR.

**Comment by mward-sudo** [comment]
@bartblast Ready for review once checks have completed / passed.

File: assets/js/erlang/filename.mjs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bartblast) [reply assets/js/erlang/filename.mjs]: @mward-sudo Yes, I agree with CodeRabbit's conclusion. The key insight is that Elixir's interop layer converts the result: when you call `:filename.rootname([], ".erl")` from Elixir, it returns `""` (binary), not `[]` (list) like pure Erlang does. This happens because Elixir prefers binaries over charlists for string handling, so the interop layer sometimes normalizes Erlang's list-based returns to binaries. Since our goal is ExJs consistency (matching Elixir's actual behavior), returning `Type.bitstring("")` at line 120 is correct.

**Comment by mward-sudo** [comment]
For `rootname/1`, `"iolist with invalid UTF-8 bytes"` is covered by `"handles invalid UTF-8 bytewise"` and related tests are more comprehensive.

I have implemented all the other suggested tests.

File: test/elixir/hologram/ex_js_consistency/erlang/filename_test.exs [hunk 3]
```diff
@@ -477,6 +477,291 @@ defmodule Hologram.ExJsConsistency.Erlang.FilenameTest do
...
+      assert :filename.rootname("foo.erl") == "foo"
+    end
+
+    test "filename without extension" do
+      assert :filename.rootname("foo") == "foo"
+    end
+
+    test "filename starting with dot" do
+      assert :filename.rootname(".foo") == ""
+    end
```
[the remark this answers is not in the document]
Comment (mward-sudo) [reply test/elixir/hologram/ex_js_consistency/erlang/filename_test.exs:499]:
```
iex(1)> :filename.rootname(".foo")
""
```

In erl:

```
1> filename:rootname(".foo").
[]
```

**Comment by bartblast** [comment]
Looks good, thanks @mward-sudo 💜
